### PR TITLE
Sending emails, part 1

### PR DESCRIPTION
**Pull request by ro-tex**
# PULL REQUEST

## Overview

This PR implements sending email and it's a first in a series that implements email communication with users.

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [x] All new methods or updated methods have clear docstrings
 - [x] Testing added or updated for new methods
 - [x] Verify if any changes impact the WebPortal Health Checks
 - [x] Approriate documentation updated
 - [x] Changelog file created

### CODE REVIEW

File: database/email.go
Comment (MSevey) [inline database/email.go:85]: Should this error be combined with the normal return error?

File: database/email.go
Comment (MSevey) [inline database/email.go:115]: same question on combining errors. Otherwise a comment explaining why not

File: database/email.go
Comment (MSevey) [inline database/email.go:120]: This could be refactored to a method that returns both the ids and the messages.

File: email/sender.go
Comment (MSevey) [inline email/sender.go:22]: Can we reference the value in the env file for consistency?

File: email/mailer.go
Comment (MSevey) [inline email/mailer.go:43]: Should these be in the env file? Thinking about how community portal operators should be updating these.

File: database/email.go
Comment (ro-tex) [reply database/email.go:85]: I don't think so. If we do that we won't send the messages we've already locked. The error is somewhat important to us as the team responsible for the infrastructure but it's not important for the process of sending emails. That's why it's logged but otherwise ignored.

File: database/email.go
Comment (ro-tex) [reply database/email.go:115]: I can add a comment.

Comment (ro-tex) [reply database/email.go:115]: Combined.

File: database/email.go
Comment (ro-tex) [reply database/email.go:120]: The reason for the current approach is that I don't have a way to lock a limited number of records, so I want to lock them by their ids. Which means that another server doing the same thing at the same time might decide to update the same records. If that happens and `fetchUnlockedMessageIDs` returns the records instead of the ids, those emails will get sent twice. With the current setup the messages updated by both servers will be sent just once and one of the servers will send fewer emails in its current batch, which is not an issue.

There might be something I'm missing here. It will be easier to discuss that on Discord.

File: email/sender.go
Comment (ro-tex) [reply email/sender.go:22]: It's already there but I haven't mentioned it in the docstring. I guess I added it to the env file later.
Updating the docsting.

File: email/mailer.go
Comment (ro-tex) [reply email/mailer.go:43]: They are. I forgot to mention it in the docstrings.
Good catch, thanks. :)

**Comment by ro-tex** [comment]
The PR is now updated to target `main`.

File: database/email.go
Comment (ChrisSchinnerl) [reply database/email.go:55]: I used this in skyd btw. Saved me a lot of work https://github.com/square/mongo-lock

File: database/email.go
Comment (MSevey) [reply database/email.go:119]: You have the same for loop that iterates over `for c.Next(ctx)` twice https://github.com/SkynetLabs/skynet-accounts/pull/72/files/7f099e6c21a8226a5ccc9f0155e67b09d5b46e06#diff-23fbde68a3027d369c3bf5fcd43e6d5cce2830e5108303066a7d9fc14e0c1358R90

Both for loops decode the EmailMessage. One loop creates a slice of EmailMessages, and one loop creates a slice of ObjectIDs. 

You could just have one helper method that returns both. The looping of the `c.Next(ctx)` feels like something that might happen for a number of reasons so might be worth having pulled out. 

Not blocking, just a thought.

File: email/sender.go
Comment (MSevey) [inline email/sender.go:99]: This can just be added to the select statement

```
s.scanAndSend()
select {
case <-s.staticCtx.Done():
case <-time.After(sleepBetweenScans):
}
```

Comment (MSevey) [reply email/sender.go:99]: Or we could use a `time.AfterFunc` here

Comment (MSevey) [reply email/sender.go:99]: nit / follow up

**Review by MSevey — APPROVED** [review]

File: database/email.go
Comment (ro-tex) [reply database/email.go:119]: I think I misunderstood the original idea. Moving this to a helper makes sense. Will add it in the next PR.

File: email/sender.go
Comment (ro-tex) [reply email/sender.go:99]: Sure thing. Changed in the upcoming PR
```go
		for {
			select {
			case <-s.staticCtx.Done():
				return
			case <-time.After(sleepBetweenScans):
				s.scanAndSend()
			}
		}
```

File: database/email.go
Comment (ro-tex) [reply database/email.go:41]: Changed.

File: database/email.go
Comment (ro-tex) [reply database/email.go]: Added.

File: database/email.go
Comment (ro-tex) [reply database/email.go]: Done.

**Review by MSevey — DISMISSED** [review]
As a follow up, is there a chance the `abortTxn` flow can be moved to a defer statement?

**Comment by ro-tex** [comment]
> As a follow up, is there a chance the `abortTxn` flow can be moved to a defer statement?

Yes! This was my right-before-falling-asleep thought last night as well, ha ha.

File: api/handlers.go
Comment (ChrisSchinnerl) [reply api/handlers.go:823]: This is also very easy to unit test since you can just add a dummy handler which you wrap with `WithDBSession` that only writes a single thing to the db and then returns an error one time and a success the other time.
So you just check whether the txn happened or not and then you can trust that as long as you pass the right context to the db, all modifications within a handler happen atomically.

Comment (ro-tex) [reply api/handlers.go]: Added.

File: database/database.go
Comment (ro-tex) [inline database/database.go:55]: This change is required for transactions. It's unfortunate but if we want transactions, we need to read from the primary.

File: api/mongowriter.go
Comment (ro-tex) [reply api/mongowriter.go:40]: Yes. :)

File: email/sender.go
Comment (ro-tex) [reply email/sender.go:149]: Makes sense.

File: database/email.go
Comment (peterjan) [inline database/email.go:73]: should we ensure we don't overwrite `sent_at` by adding `sent_at: null` in the predicate?

File: database/email.go
Comment (peterjan) [inline database/email.go:100]: this is a pretty weird check right, shouldn't it be `m.FailedAttemptsToSend == maxAttempts`

File: email/sender.go
Comment (peterjan) [inline email/sender.go:132]: needs an index here as well

File: email/sender.go
Comment (peterjan) [inline email/sender.go:214]: `sendMultiple` docstring mismatch

File: test/fixtures/mongo_keyfile
Comment (peterjan) [inline test/fixtures/mongo_keyfile:1]: this is safe to commit right?

File: email/sender.go
Comment (peterjan) [inline email/sender.go:33]: seems high

File: database/email.go
Comment (peterjan) [inline database/email.go:93]: why are you collecting `ids`?

File: database/email.go
Comment (peterjan) [inline database/email.go:22]: why are some fields `omitempty`?

File: database/email.go
Comment (peterjan) [inline database/email.go:122]: could've avoided having two fields by having `failed_attempts` and init it at zero, and then do `failed: {$lt: maxFailedAttempts}` in a predicate that fetches emails that need to be retried

**Review by peterjan — CHANGES_REQUESTED** [review]
only requesting changes for the missing index, otherwise it LGTM

File: Makefile
Comment (ro-tex) [reply Makefile:76]: Yes. This thing is spinning up a test container that's exposing its port on the host system. If the host system is running another mongo container (as mine is 90% of the time) this will fail for a silly and mundane reason. So, I switched it to a custom port in order to avoid that.

File: api/cookie.go
Comment (ro-tex) [reply api/cookie.go:36]: We can but it doesn't really matter. We'll take what we need even if there is more data there. This makes it OK for you to test by just holding the `A` key down for a bit without counting characters.

File: database/email.go
Comment (ro-tex) [reply database/email.go:73]: If the message id reaches this line it means it was just sent. In the worst case we will update a `sent_at` value that's milliseconds before the current one. The real issue here would be the fact that an email was sent multiple times. And even that is not a real problem. We definitely want to avoid it but even if it happens, it's not an issue.

File: database/email.go
Comment (ro-tex) [reply database/email.go:100]: The reason for this check is the fact that we increment the field in the next DB call. If I increment in one call and then check the value and mark the message as failed in another, we would have what you describe. But if we don't want to update records twice (and either re-fetch them from the DB or scan the entire collection) we need to do it this way.

File: email/sender.go
Comment (ro-tex) [reply email/sender.go:132]: Added.

File: test/fixtures/mongo_keyfile
Comment (ro-tex) [reply test/fixtures/mongo_keyfile:1]: Yes. This is not our key file, it's the key file we use to connect to the test mongodb replica set which always starts empty.

File: email/sender.go
Comment (ro-tex) [reply email/sender.go:33]: What would be a more adequate value? 1 second?

File: database/email.go
Comment (ro-tex) [reply database/email.go:93]: Oh, nice catch! Dead code. I was using them for locking before.

File: database/email.go
Comment (ro-tex) [reply database/email.go:22]: If you don't set the `_id` filed to `omitempty` you need to generate the ID yourself, otherwise the record gets inserted in the DB with a `00000000000000` ID which is not what we want. The same is true for `sent_at` where we won't be able to check if it's `nil` in the query but we'll need to compare it to a specific zero value.

File: database/email.go
Comment (ro-tex) [reply database/email.go:122]: Good idea.

**Review by peterjan — APPROVED** [review]
LGTM 👍

**Review by ChrisSchinnerl — APPROVED** [review]

**Pull request merged**
